modules/service/src/main/scala/lucuma/odb/data/ProgramReference.scala

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.data
+
+import cats.Order
+import cats.parse.Numbers.digit
+import cats.parse.Parser
+import cats.parse.Parser.*
+import eu.timepit.refined.cats.given
+import eu.timepit.refined.types.numeric.PosInt
+import io.circe.Decoder
+import io.circe.Encoder
+import io.circe.syntax.*
+import lucuma.core.enums.Half
+import lucuma.core.model.Semester
+import lucuma.core.optics.Format
+import lucuma.core.parser.MiscParsers.intN
+import lucuma.core.parser.MiscParsers.posInt
+
+import java.time.DateTimeException
+import java.time.Year
+import scala.util.control.Exception
+
+case class ProgramReference(semester: Semester, index: PosInt) {
```

Comment:
As mentioned, this is missing the program type and will probably become a `trait` with implementations for the various forms: science vs `CAL` / `ENG` / `LIB` / `XPL`.  Then I suppose it will go into `core` along with updates to `Semester` and its parsing.